codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3UpdateEndpoint.java

```diff
@@ -0,0 +1,154 @@
+import software.amazon.smithy.go.codegen.integration.ConfigField;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.go.codegen.integration.MiddlewareRegistrar;
+import software.amazon.smithy.go.codegen.integration.RuntimeClientPlugin;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.utils.ListUtils;
+
+public class S3UpdateEndpoint implements GoIntegration {
+    private static final String USE_PATH_STYLE_OPTION = "UsePathStyle";
+    private static final String USE_ACCELERATE_OPTION = "UseAccelerate";
+    private static final String UPDATE_ENDPOINT_ADDER = "addUpdateEndpointMiddleware";
+    private static final String UPDATE_ENDPOINT_INTERNAL_ADDER = "UpdateEndpoint";
+
+    /**
+     * Gets the sort order of the customization from -128 to 127, with lowest
+     * executed first.
+     *
+     * @return Returns the sort order, defaults to -40.
+     */
+    @Override
+    public byte getOrder() {
+        return 127;
+    }
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoDelegator goDelegator
+    ) {
+        if (!isS3Service(model, settings.getService(model))) {
+            return;
+        }
+
+        goDelegator.useShapeWriter(settings.getService(model), this::writeMiddlewareHelper);
+
+        // Generate getter's for an operation input that can be used to satisfy interfaces
+        for (ShapeId shapeId: settings.getService(model).getOperations()) {
+            OperationShape operation = model.expectShape(shapeId, OperationShape.class);
+            StructureShape input = model.expectShape(operation.getInput().get(), StructureShape.class);
+            goDelegator.useShapeWriter(operation, writer -> {
+                writeInputGetter(writer, symbolProvider, input);
+            });
+        }
+    }
+
+    private void writeInputGetter(GoWriter writer, SymbolProvider symbolProvider, StructureShape input) {
+        // generate bucketGetter if input has a member named Bucket
+        if (input.getMember("Bucket").isPresent()) {
+            // generateBucketGetter
+            // TODO: Look for alternatives so that these Getters are NOT exported?
+            Symbol inputSymbol = symbolProvider.toSymbol(input);
+            writer.writeDocs("GetBucket retrieves the Bucket member value if provided");
+            writer.openBlock("func (s $P) GetBucket() (v string) {", "}", inputSymbol, () -> {
+                writer.write("if s.Bucket == nil { return v }");
+                writer.write("return *s.Bucket");
+            });
+            writer.insertTrailingNewline();
+        }
+    }
```

Comment:
The v1 SDK used the getter pattern since the types, client, and customizations were all in the same package so they could be private. V2 Doesn't have this benefit, and would need to export any helper methods.

Because of this, i think we want to look at a different design strategy for the SDK's middleware customizations to get access to the API parameter's members. This strategy would be similar to how operation middleware is generated per operation. Similar idea could be applied here. 

The generator could generate a helper function to extract the `BucketName` member from a relevant input type.  This helper could either be generated per operation, or shared across multiple operations using a type switch to selectively grab the inputs.  Generating the member helpers per operation probably is easier from a type point of view, but probably more complicated when looking at this from the middleware point of view.

From the middleware point of view it would make sense to have a "generic" generated accessor function for multiple input types that knows the member of the specific input type to extract the value from.

With this design change, no additional methods need to be generated on the types, and all accessor helpers stay private and only used internally by the middleware helper that is generated.